### ISSUE TITLE
Cannot find module phoenix\priv\static\phoenix.js\package.json on startup, on windows

### ISSUE THREAD

**Issue by Sushisource** [phoenixframework/phoenix#1463]
After a fresh install of the latest phoenix (1.1.2)  and the latest node/npm (5.4.0/3.5.3) while creating a brand new project, and using the latest erlang (although I doubt that matters for this bug) I immediately encounter this:

    13:29:58.10 E:\eci>mix phoenix.new .
    * creating ./config/config.exs
    ..... yada yada yada ....

    Fetch and install dependencies? [Yn] y
    * running mix deps.get
    * running npm install && node node_modules/brunch/bin/brunch build
    npm WARN install Couldn't install optional dependency: Unsupported
    npm WARN prefer global coffee-script@1.10.0 should be installed with -g
    npm WARN EPACKAGEJSON eci No license field.
    module.js:328
        throw err;
        ^

    Error: Cannot find module 'E:\eci\node_modules\phoenix_html\priv\static\phoenix_html.js\package.json'
      at Function.Module._resolveFilename (module.js:326:15)
      at Function.Module._load (module.js:277:25)
      at Module.require (module.js:354:17)
      at require (internal/module.js:12:17)
      at filePath (E:\eci\node_modules\brunch\lib\deppack.js:53:33)
      at filePath (E:\eci\node_modules\brunch\lib\deppack.js:56:37)
      at Object.keys.filter.forEach (E:\eci\node_modules\brunch\lib\deppack.js:71:17)
      at generateModule (E:\eci\node_modules\brunch\lib\deppack.js:99:24)
      at E:\eci\node_modules\brunch\lib\deppack.js:151:13
      at tryToString (fs.js:414:3)
      at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:401:12)

    We are all set! Run your Phoenix application:
    ...etc...

This is on Windows 7.

As a result, the static assets aren't compiled, and the welcome page has no styling. Let me know if I can provide additional details to help the fix. Node/npm is a frequent headache on windows so I'm not too surprised it's being obnoxious.

**Comment by chrismccord** [comment]
Can you include the entire output of `npm install` when run inside the new project's directory?

**Comment by chrismccord** [comment]
Can you also include your package.json and mix.lock ?

**Comment by Sushisource** [comment]
Here's from npm install:

    14:26:49.05 E:\eci>npm install
    npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
    npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.6
    npm WARN eci No license field.

Here's a zip of the two files
[mix_package.zip](https://github.com/phoenixframework/phoenix/files/83353/mix_package.zip)

The project is totally unmodified. Thanks for the quick response!

**Comment by josevalim** [comment]
@Sushisource can you try changing those entries in your `package.json` to the following:

```json
"phoenix": "file:deps\\phoenix",
"phoenix_html": "file:deps\\phoenix_html"
```

And run `npm install` again? I am wondering if it is a slash thing. Also can you check that both deps/phoenix and deps/phoenix_html exist? Thank you!

**Comment by chrismccord** [comment]
wait so the 2nd npm install worked properly?  Are you sure you ran it in the project directory? If you did, start up the server and see if it works

**Comment by josevalim** [comment]
@chrismccord Ah, I see, the second one worked didn't see to have an error report. :D

**Comment by Sushisource** [comment]
It's not the npm install that fails, rather the running of brunch is what fails. Example:

    14:32:27.79 E:\eci>node_modules\.bin\brunch.cmd  build
    module.js:328
        throw err;
        ^

    Error: Cannot find module 'E:\eci\node_modules\phoenix\priv\static\phoenix.js\package.json'
      at Function.Module._resolveFilename (module.js:326:15)
      at Function.Module._load (module.js:277:25)
      at Module.require (module.js:354:17)
      at require (internal/module.js:12:17)
      at filePath (E:\eci\node_modules\brunch\lib\deppack.js:53:33)
      at filePath (E:\eci\node_modules\brunch\lib\deppack.js:56:37)
      at Object.keys.filter.forEach (E:\eci\node_modules\brunch\lib\deppack.js:71:17)
      at generateModule (E:\eci\node_modules\brunch\lib\deppack.js:99:24)
      at E:\eci\node_modules\brunch\lib\deppack.js:151:13
      at tryToString (fs.js:414:3)
      at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:401:12)

**Comment by Sushisource** [comment]
@josevalim Changing the paths like that doesn't seem to affect it (ran npm install, then brunch). Which makes sense considering Windows does support forward slashes (although quite a few tools can mess it up)

**Comment by josevalim** [comment]
I see, it is a brunch error. One last shot: can you access deps/phoenix/package.json and deps/phoenix_html/package.json and prefix their "index" key by "./"?

**Comment by chrismccord** [comment]
Paging @paulmillr (sorry!). Any quick ideas on if this is a brunch bug or further downstream in a brunch dep?

**Comment by josevalim** [comment]
To be clear: prefix the "main" key by "./". So it should be something like "./priv/static/phoenix.js".

**Comment by Sushisource** [comment]
@josevalim Looks like it already is:

    {
      "name": "phoenix_html",
      "version": "2.3.1",
      "main": "./priv/static/phoenix_html.js",
      "repository": {
      },
      "devDependencies": {
        "brunch": "^2.1.1",
        "babel-brunch": "^6.0.0",
        "uglify-js-brunch": ">= 1.0 < 1.8"
      }
    }

**Comment by josevalim** [comment]
Ah, they seem to be normalized already then. :D Ok, I am ran out of ideas. If you feel a little bit adventurous, you can open up the brunch dependency here:

```
at filePath (E:\eci\node_modules\brunch\lib\deppack.js:53:33)
at filePath (E:\eci\node_modules\brunch\lib\deppack.js:56:37)
```

At some point somewhere it seems the paths are not being joined as they were supposed to. I will output what mine look like in a later comment.

**Comment by Sushisource** [comment]
AFAICT it looks like the entire 

    priv\static\phoenix.js

Section shouldn't be in that path... I poked around in the bits of brunch you mentioned but it looks like it's probably gone wrong earlier in the stack.

I've got to run for the rest of the day, but I'll check back tomorrow. Thanks so much for looking into this. Loving what you guys are doing with elixir/phoenix and I'm super impressed that you guys care about windows at all, which seems rare these days.

**Comment by josevalim** [comment]
I think I may know the root cause. There is a function called `getModuleRootPath` in `deppack.js`. If it receives a path with "\\", it is going to work incorrectly and return the given path, instead of finding the root. Here is its implementation:

```javascript
const getModuleRootPath = path => {
  const split = path.split('/');
  const index = split.lastIndexOf('node_modules');
  return split.slice(0, index + 2).join('/');
};
```

You can be sure there is a bug by logging the input with `console.log(path)`. If the path you receive contains the `\\`, then it will no longer work. You can fix it by changing `getModuleRootPath` and the next function to use `path.sep` instead of `'/'`:

```javascript
const getModuleRootPath = path => {
  const split = path.split(path.sep);
  const index = split.lastIndexOf('node_modules');
  return split.slice(0, index + 2).join(path.sep);
};
```

Please let us know if this happen to the root cause.

**Comment by bendennis** [comment]
@josevalim I added the log output to getModuleRootPath in deppack.js as you suggested and am not seeing the ``` \\```. When I start the server I see the following output, including the "PATH - " console log statement:
```
$ mix phoenix.server
[info] Running Blog.Endpoint with Cowboy using http on port 4000
PATH - c:\Users\Ben\workspace\phoenix\blog\node_modules\phoenix\priv\static\phoenix.js
module.js:328
    throw err;
    ^

Error: Cannot find module 'c:\Users\Ben\workspace\phoenix\blog\node_modules\phoenix\priv\static\phoenix.js\package.json'
  at Function.Module._resolveFilename (module.js:326:15)
  at Function.Module._load (module.js:277:25)
  at Module.require (module.js:354:17)
  at require (internal/module.js:12:17)
  at filePath (c:\Users\Ben\workspace\phoenix\blog\node_modules\brunch\lib\deppack.js:54:33)
  at filePath (c:\Users\Ben\workspace\phoenix\blog\node_modules\brunch\lib\deppack.js:57:37)
  at Object.keys.filter.forEach (c:\Users\Ben\workspace\phoenix\blog\node_modules\brunch\lib\deppack.js:72:17)
  at generateModule (c:\Users\Ben\workspace\phoenix\blog\node_modules\brunch\lib\deppack.js:100:24)
  at c:\Users\Ben\workspace\phoenix\blog\node_modules\brunch\lib\deppack.js:152:13
  at tryToString (fs.js:414:3)
  at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:401:12)
```

**Comment by josevalim** [comment]
You are seeing the bug! Notice your path contains `\` instead of `/`. So the `path.split("/")` will never work. You can fix by replacing that function and the next one by the following:

```javascript
const getModuleRootPath = path => {
  const split = path.split(path.sep);
  const index = split.lastIndexOf('node_modules');
  return split.slice(0, index + 2).join(path.sep);
};

const getModuleRootName = path => {
  const split = path.split(path.sep);
  const index = split.lastIndexOf('node_modules');
  return split[index + 1];
};
```

Please do those changes and let us know if it works. I will send brunch a PR meanwhile.

**Comment by bendennis** [comment]
Oops, I edited my post above instead of making a new one after I figured out what was going on last night. Should have just responded in line.

Nevertheless, I did change the split functions as you suggested and all is working now!

Thank you!

**Comment by Sushisource** [comment]
Indeed, same here. Added the logging statement and got this as output:
      
    E:\eci\node_modules\phoenix\priv\static\phoenix.js

 which still makes me think the entire

    priv\static\phoenix.js

Is getting in there when it shouldn't be

**Comment by josevalim** [comment]
@Sushisource can you make sure you are using brunch 6.1.3? This should be fixed on latest brunch. :)

**Comment by Sushisource** [comment]
Yup, it's all fixed now on brunch 2.1.3. Thanks much @josevalim
